### ISSUE TITLE
What exactly does it mean that "Function calls are not supported" in AOT metadata

### ISSUE THREAD

**Issue by elca-lan** [angular/angular#45753]
### Description

This is more of a request for clarification, but depending on the outcome, the documentation could need some updates. The thing that I don't understand is the section [Function calls are not supported](https://angular.io/guide/aot-metadata-errors#function-calls-not-supported).

First of all, I'm reasonably sure that the title is misleading and/or outdated since function **calls** do very much seem to be allowed in metadata / decorators, since they are commonly used in Angular code and documentation. For example, the documentation of [RouterModule](https://angular.io/api/router/RouterModule) contains the common example of
```
@NgModule({
  imports: [RouterModule.forChild(ROUTES)]
})
class MyNgModule {}
```
which according to my understanding constitutes a function call (`RouterModule.forChild(...)`) and is part of a decorator and therefore metadata.

Then, the first actual statement in that documentation section is "The compiler does not currently support [function expressions or lambda functions](https://angular.io/guide/aot-compiler#function-expression)" which again, gives me reason to pause since that is something completely different than function **calls**.

But even if that's just a confusion that is specific to me, then there is still the fact that everything that this section describes as generating errors, seems to work just fine _with Ivy_. I've been using older, pre-Ivy versions of Angular for a long time and am generally familiar with the described issues, but as far as I can tell, none of that applies to modern Angular applications anymore.

Is there a bug in my _interpretation_ of the documentation? What is the _context_ that would make my examples fail?

I've been (mostly) sticking to the "approved" patterns in my code, but since it causes significant overhead and is counter intuitive to newcomers, I would like to know how it actually works (or doesn't work).

### What is the affected URL?

https://angular.io/guide/aot-metadata-errors#function-calls-not-supported

### Please provide the steps to reproduce the issue

Simply create a new Angular project using the Angular CLI and update the content of app.component.ts to
```
import { Component, Inject, InjectionToken } from '@angular/core';

export const TitleToken = new InjectionToken("TitleToken");

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css'],
  providers: [
    {
      provide: TitleToken,
      useFactory: () => {
        return `Injected ArrowFactory Title @ ${new Date()}`
      }
    }
  ]
})
export class AppComponent {
  public constructor(
    @Inject(TitleToken) public readonly title: string
  ) {}
}
```

### Please provide the expected behavior vs the actual behavior you encountered

The sample above builds and runs just fine in both development an production mode and the documentation says that it should not.

**Comment by JoostK** [comment]
This is a complicated area of Angular and the gist of it is mostly as you describe already: some of the limitations have been lifted with Angular Ivy and the documentation page is indeed outdated.

Regarding function calls: only functions with a single return-statement can be evaluated, even with Ivy. This was also the case with the View Engine compiler, however Ivy _is_ able to evaluate function expressions and arrow functions.

The example with `Component.providers` is interesting, as DI metadata is no longer statically analyzed in Ivy so the AOT compilation restrictions do not apply there anymore (since Ivy).

**Comment by elca-lan** [comment]
Thanks a lot @JoostK, that is very good to know. Maybe I can help with a PR for updating that documentation?

But first, I have some follow-up questions:

> only functions with a single return-statement can be evaluated
That one I have never heard / read before, so I just tried the following example which surprisingly still works:
```
  providers: [
    {
      provide: TitleToken,
      useFactory: () => {
        if (Date.now() % 2 == 0) {
          return `Injected ArrowFactory Title @ ${new Date()}`;
        } else {
          return "ODD";
        }
      }
    }
  ]
```
I have also checked the "Destructured variable or constant not supported" limitation which also does not seem to apply. I'll go and check the rest of them too if I find some time.

> The example with Component.providers is interesting, as DI metadata ...
But isn't that _literally_ the context of that documentation? The section describes the limitations of the AOT that specifically apply to what types of expressions are allowed in decorators, where the most common case for running into those issues happens to be `Compontnt.providers`.

I understand that internally, these decorators are processed and added as "static" metadata to the class definition etc., but I don't see how that would change what is allowed and what is not. I guess I'm just still slightly confused.

So, in short: do **none** of limitations specifically mentioned in "Function calls are not supported" apply to Component/Directive/NgModule metadata anymore in Ivy?

If so, then updating the documentation would be a pretty easy job for me :)

**Comment by JoostK** [comment]
> So, in short: do none of limitations specifically mentioned in "Function calls are not supported" apply to Component/Directive/NgModule metadata anymore in Ivy?

Nope, most of the restrictions actually still apply, but not for `Component.providers`/`NgModule.providers`. But if you were to use a complex function within an `NgModule.imports` field then you'd run into trouble:

```ts
function myComplexFunction() {
  const result = [];
  result.push(ImportedModule);
  return result;
}

@NgModule({
  imports: myComplexFunction(),
})
```

The above does report an error because `NgModule.imports` must be statically analyzable. There are some more decorator fields that need to be statically analyzable, but not all of them are.

---

Your example with `useFactory` works because the Ivy compiler does not statically evaluate the list of providers; it just copies the array of providers into the compiled output verbatim. This is unlike the VE compiler, which had to record _all_ metadata in a decorator.

**Comment by elca-lan** [comment]
Wow! Mind blown 🤯.

I think I'm starting to get the hang of it. In your example, the error message say "Unable to evaluate function call of complex function. A function must have exactly one return statement.", but that is _not exactly_ correct, right? It must "consist" of exactly one return statement, so the function body needs to have the form `return ...` for the static analyzer to "get it", which makes sense because it must be able to _copy_ the method result to the metadata statically, without executing the code... at least I think that's right.

And `NgModule.imports` specifically MUST be statically analyzable because the compiler needs to figure out the dependencies between all the modules.

And the reason why having calls to functions like `RouterModule.forChild` work in `imports` is because  `ModuleWithProviders<T>` still statically declares its module dependency through `T`, right? And inside of such a function that retuns a ModuleWithProviders<T>` I can then again do pretty much whatever I want - at least all the weird things that I just tried.

There seem to be a lot of non-trivial and interesting special cases that I'll have to look into.

**Comment by JoostK** [comment]
> the error message say "Unable to evaluate function call of complex function. A function must have exactly one return statement.", but that is not exactly correct, right? It must "consist" of exactly one return statement

Interesting; I get where you're coming from as I can see how the message can be interpreted in two ways; it is meant as "must have exactly one statement which needs to be a return statement". If changing the wording to "consists" avoids this ambiguity then I'm fine with changing it to that.

> And the reason why having calls to functions like RouterModule.forChild work in imports is because ModuleWithProviders<T> still statically declares its module dependency through T, right? And inside of such a function that retuns a ModuleWithProviders<T> I can then again do pretty much whatever I want - at least all the weird things that I just tried.

Yep, you got it _exactly_ correct!
